core/peerset/src/lib.rs

```diff
@@ -47,8 +40,84 @@ struct Inner {
 	in_slots: Vec<Option<PeerId>>,
 }
 
-/// Message that can be sent by the peer set manager (PSM).
 #[derive(Debug)]
+enum Action {
+	AddReservedPeer(PeerId),
+	RemoveReservedPeer(PeerId),
+	SetReservedOnly(bool),
+	ReportPeer(PeerId, i32),
+	Incoming(PeerId, IncomingIndex),
+	Dropped(PeerId),
+	Discovered(PeerId),
+}
+
+/// Shared handle to the peer set manager (PSM). Distributed around the code.
+#[derive(Debug, Clone)]
+pub struct PeersetHandle {
+	tx: mpsc::UnboundedSender<Action>,
+}
+
+impl PeersetHandle {
+	/// Adds a new reserved peer. The peerset will make an effort to always remain connected to
+	/// this peer.
+	///
+	/// Has no effect if the node was already a reserved peer.
+	///
+	/// > **Note**: Keep in mind that the networking has to know an address for this node,
+	/// >			otherwise it will not be able to connect to it.
+	pub fn add_reserved_peer(&self, peer_id: PeerId) {
+		let _ = self.tx.unbounded_send(Action::AddReservedPeer(peer_id));
+	}
+
+	/// Remove a previously-added reserved peer.
+	///
+	/// Has no effect if the node was not a reserved peer.
+	pub fn remove_reserved_peer(&self, peer_id: PeerId) {
+		let _ = self.tx.unbounded_send(Action::RemoveReservedPeer(peer_id));
+	}
+
+	/// Sets whether or not the peerset only has connections .
+	pub fn set_reserved_only(&self, reserved: bool) {
+		let _ = self.tx.unbounded_send(Action::SetReservedOnly(reserved));
+	}
+
+	/// Reports an adjustement to the reputation of the given peer.
+	pub fn report_peer(&self, peer_id: PeerId, score_diff: i32) {
+		let _ = self.tx.unbounded_send(Action::ReportPeer(peer_id, score_diff));
+	}
+
+	/// Indicate that we received an incoming connection. Must be answered either with
+	/// a corresponding `Accept` or `Reject`, except if we were already connected to this peer.
+	///
+	/// Note that this mechanism is orthogonal to `Connect`/`Drop`. Accepting an incoming
+	/// connection implicitely means `Accept`, but incoming connections aren't cancelled by
+	/// `dropped`.
+	///
+	/// Because of concurrency issues, it is acceptable to call `incoming` with a `PeerId` the
+	/// peerset is already connected to, in which case it must not answer.
+	pub fn incoming(&self, peer_id: PeerId, index: IncomingIndex) {
```

Comment:
The `incoming`, `dropped` and `discovered` methods were only on the `PeersetMut` (now `Peerset`) because it doesn't make sense for them to be accessible from anything else than the networking. If someone calls them you will just crate an inconsistent state in the network layer. If for some reason we have to keep that as is, please put a huge warning that says "don't call these functions".

Reply:
Why would anyone want to call this functions? Calling them is a logical error, same as doing

```rust
let a = None;
if let Some(value) = a {
   // ... 
}
``` 

`Peerset` is a `futures::Stream` and I believe it should be responsible only for producing items, nothing else. 

If we want to limit how and where the methods are used, I believe we should have 2 different handles. One with `incoming`, `dropped` and `discovered` methods and the second one with everything else

Reply:
> Calling them is a logical error

Yeah, and it would be even better if the type system prevents these logic errors.

> Peerset is a futures::Stream and I believe it should be responsible only for producing items, nothing else.

Why is that?

Reply:
>> Peerset is a futures::Stream and I believe it should be responsible only for producing items, nothing else.
>
> Why is that?

So we can use combinators (`.map`, `.then`) on the stream

Reply:
The only thing that will use the `Peerset` is the networking code which already exists and doesn't use combinators. I don't think we should have a weird API and add a performance hit because of usability, especially since usability here is a non-issue.